pytorch_lightning/plugins/checkpoint/checkpoint.py

```diff
@@ -0,0 +1,28 @@
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Dict, Optional, Union
+
+
+class CheckpointIOPlugin(ABC):
+    @abstractmethod
+    def save_checkpoint(
+        self, checkpoint: Dict[str, Any], path: Union[str, Path], storage_options: Optional[Any] = None
+    ) -> None:
+        """Save model/training states as a checkpoint file through state-dump and file-write.
+
+        Args:
+            checkpoint: dict containing model and trainer state
+            path: write-target path
+            storage_options: Optional parameters when saving the model/training states.
+        """
+
+    @abstractmethod
+    def load_checkpoint(self, path: Union[str, Path], storage_options: Optional[Any] = None) -> Dict[str, Any]:
+        """
```

Comment:
We should type the `storage_options` with TypeVar instead of Any
otherwise all implementations with deal with their storage_options being more restrictive than the base class, which will upset typecheckers

```
TSaveStorageOptions = TypeVar("S")
TLoadStorageOptions = TypeVar("T")

@abstractmethod
    def save_checkpoint(
        self, checkpoint: Dict[str, Any], path: Union[str, Path], storage_options: Optional[S] = None
    ) -> None:

@abstractmethod
    def load_checkpoint(self, path: Union[str, Path], storage_options: Optional[T] = None) -> Dict[str, Any]:
```

we did this for broadast here: https://github.com/PyTorchLightning/pytorch-lightning/blob/cb2a8ed1b8f78386de4771d3bc59e19022e1b023/pytorch_lightning/plugins/training_type/training_type_plugin.py#L102-L104

Reply:
Awesome, thanks @ananthsub :)

Reply:
@ananthsub Why do we need to enforce this?
there's no way this abstract interface knows the needs of its children.

And the TypeVars doesn't seem to be linked across input/outputs or methods.

IMO storage_options should be `**storage_options: Any` or `storage_options: Optional[Any] = None`